### ISSUE TITLE
fix-unit-occupancy: regular check too slow for large fortresses: optimise and/or doc warning?

### ISSUE THREAD

**Issue by TheBloke** [DFHack/dfhack#749]
I have found that I cannot run ```fix-unit-occupancy enable``` on my largest fortress without unacceptable freezes.

This fort is 3x3 / 144x144 tiles.  I have dug down to Z -40 and all caverns are visible (though not pathable by dwarves.)  All Z levels from +15 down to -9 are 100% visible, and below that I would estimate an average of 30% of each level is revealed.  There are 268 dwarves which also contributes to the slowdown, though I suspect less so than the amount of visible map. Save file available on request.

On this fort, each time fix-unit-occupancy runs it takes approximately 6.3 seconds.  The default timer of 1200 ticks equates to 40 seconds at my current 30 FPS.  So I have 6.3 seconds of slowdown every 40 seconds of gameplay - a total loss of about 15% of game time (6.5 seconds frozen per 46.5 seconds elapsed.)

Examples from my stderr.log:
```fix-unit-occupancy: elapsed time: 6.362 secs```
```fix-unit-occupancy: elapsed time: 6.37598 secs```
```fix-unit-occupancy: elapsed time: 6.367 secs```

I have also tested it on my second fort, which is only 18 game-months old and has very significantly less dug ground.  This fort gets 120 max FPS and I can detect no lag from running fix-unit-occupancy.

I thought I should raise an issue:
* To see if there are performance improvements that can be made?  Is it possible to scan for the bug without freezing DF the whole time?  Running the scan from a separate thread?
  * Maybe the scan can even occur exactly as it does now, but in another thread?  Can ```world->map.map_blocks``` and ```world->units.active```  be accessed live from a second thread while DF remains running?  If so, that might be a simple with minimal rewrite.
  * If not, what about taking a copy of ```world->map.map_blocks``` and ```world->units.active``` and scanning them in the second thread?
     * ie DF is frozen only for the duration of copying ```map_blocks``` and ```units.active```, then scanning occurs in a background thread; DF is frozen again only if game changes are needed on a few blocks at the end.
    * I don't know the cost of such a copy but presumably not all elements of these objects would need to be copied, and given the bug scan has to loop 256 times for each block, checking units.active on each iteration, it seems likely that a copy and offline scan should be faster?
    * And perhaps fast copies can be done of these objects: a memcpy sort of thing?  
    * The total processing time will increase, but as the majority of it would be occurring in the background with DF still running it should be largely irrelevant to the user (even single-core users will prefer a slowdown of  a running DF to a completely frozen DF.)
    * Given the nature of the bug it seems to me that it shouldn't matter that the copies are out of date while being scanned: the plugin only changes blocks found to be bugged, and to be safe it can re-check these before applying the fix; actual bugged blocks will represent a minuscule fraction of total runtime.
* Or, if there's nothing that can be done to optimise its basic function, perhaps the scan could be (optionally) segmented?  Eg only scan x% of the map each time for 1/x successive runs.
  * In my example, if 10% of the blocks were scanned each time, each run would be 0.6 seconds which might be bearable.  
* Finally, if no optimisations are feasible, or until such time as they can be implemented, I think there should be a documentation warning indicating the potential problem.
  * Warning of possible delay is important because the slowdown might not be noticeable for some time: if the user first uses the command on a small fort, they will have no problem.  But over time as the fort grows the delay will become increasingly noticeable, and this might then be hard to track down as they will have made no recent changes to config.
  * Admittedly it does print a clear timing output to stderr.log, which *hopefully* would be seen - but it's not guaranteed for all users.  Worst case, it's the sort of creeping slowdown that could cause some users to disable DFHack completely, or at least to get the impression that it's slow.  (I realise that it's not on by default so the risk is probably minimised; but it's possible a pack might choose to distribute with it enabled.)

**Comment by lethosor** [comment]
It could definitely be optimized. For comparison, how long does reveal take? (Also, does fix-unit-occupancy take longer after revealing everything? I thought I only had it check revealed blocks, but that could be a major improvement if not.)

**Comment by TheBloke** [comment]
It does only check revealed blocks(tiles?), but this runs inside the 16x16 inner loops, so it still loops 256 times for each ```world->map.map_blocks```, but  this stops it checking occupancy versus ```world->units.active```   : 

```C++
    if (block->designation[x][y].bits.hidden)
                    continue;
```

reveal is basically instantaneous!

**Comment by TheBloke** [comment]
After running reveal, fix-unit-occupancy time increases from 6.3s to 8.9s

**Comment by TheBloke** [comment]
I can see that reveal does the same 16x16 loops for each map_block, so that part must be super fast.

Maybe it's the check of each tile against ```world->units.active``` that's so slow?  I can test again with the same map but fewer dwarves.

**Comment by TheBloke** [comment]
Interesting.  With total number of alive dwarves reduced from 268 to 193  (27% reduction), total time for fix-unit-occupancy reduces by only 0.05s!  From 6.0s to 5.95s, barely anything.

So it's not the time it takes to loop through map blocks (which reveal does instantaneously), and it doesn't seem to be the unit checking.

I can't see any other calls that look like they might be major time sinks, but I don't know enough about everything to be sure.   I'll put some timing logging in to see exactly where the time is spent.

**Comment by lethosor** [comment]
units.active includes all wildlife as well.

**Comment by TheBloke** [comment]
Ah, good to know.   I have zero pets and tame animals in this save (I slaughtered them en-masse to save FPS for some other testing), but I suppose there must be a few animals in the wild.

Actually, 24 according to Units->Others.

So it was around 6.0s with 268 + 24 = 292 units, versus 5.95s with 193 + 24 = 217 units, a 25% reduction in unit count but a basically negligible performance improvement.

Which is odd because, as you probably already know, findUnit() takes all the time:

fix-unit-occupancy: elapsed time: 10.1065 secs
fix-unit-occupancy: time hidden: 0.164711 secs
fix-unit-occupancy: time findUnit: 9.05713 secs
fix-unit-occupancy: time cur_occup: 0.148844 secs

('hidden' and 'cur_occup' are all the rest of the code that occur before and after the findUnit() call respectively, in the inner loop)

**Comment by warmist** [comment]
I would disallow (or at least discourage) this to be run periodically. There is 0 reasons why it must be done. Instead we should focus on scripts and plugins that mess up the occupancy in the first place and add to the docs: if you think that something is messed up (e.g. can't build in empty place) try running this.

**Comment by TheBloke** [comment]
Last night lethosor fixed the performance so completely that it now runs 2000 times faster :)  0.004 seconds on my large fort, compared to 8 seconds before.

So performance is no longer a reason to not run it.  But @warmist, you are saying it's not a good thing run anyway?

I must admit I possibly misunderstood what it does.  When I first saw the new feature, and looked at the bug report, I thought this was a DF bug that could happen in the normal game.  But then when I tested it last night for lethosor, he showed me that it can happen when scripts, like gm-editor, do a bad edit.

So is this a bug that only happens when DFHack-type scripts are used?  Or can it happen from a normal DF game?

If the former, then I agree there is no reason for 99% of users to run this.  If the latter, then it does sound like a fix that could be run in the background,  just in case.

Either way: is there a *risk* to running it regularly?  It seems quite safe to me, only fixing tiles that are definitely broken?  But of course I don't yet understand all the systems and repercussions.

In any case, at least performance is no longer a problem at all :)

**Comment by warmist** [comment]
There is zero risk (or at least close to zero). But the idea itself is flawed: it encourages script makers to make bad scripts and just hope that this will fix it.
Also there is the idea of "death by a thousand cuts". Meaning that each thing that you add to happen (no matter how fast) periodically the program itself slows down and the sum of a lot (there is already many dfhack things that run periodically - even each frame) such plugins/scripts makes game slow. What is worse: there is no way to optimise such slowness because it's just a lot of fast things that just add up.
Tl;dr: let's not encourage doing stuff that is not needed and guide script/plugin makers somewhat

**Comment by TheBloke** [comment]
OK thanks.  I had thought that it affected the vanilla game, I did not realise it only applied to issues caused by scripts/plugins.  

It is already not enabled by default.  Maybe just an update to the documentation, indicating this is only necessary in rare instances to fix problems caused by other scripts/plugins, and not for the vanilla game?

**Comment by lethosor** [comment]
There are occupancy bugs that can occur in vanilla DF - I've encountered at least one without using DFHack. I'm not trying to encourage people to make tools that don't work with occupancy flags properly, but I figured it should support some type of repeat functionality like fix/stuck-doors.

**Comment by lethosor** [comment]
http://www.bay12games.com/dwarves/mantisbt/view.php?id=3499 is one, along with with four duplicates reported - I think one was what encouraged me to write this plugin in the first place.

**Comment by TheBloke** [comment]
OK great, then I'm definitely going to have it running always then.   Performance is now absolutely negligible and it seems pretty low risk.

I do understand in principle warmist's concern about not wanting to make it common to band-aid fix the effects of bad code, but when a bugfix applies to a DF bug as well as mod/script bugs, then I don't think it can be seen as encouraging bad code.   Besides, it's not like it's on by default and everyone is going to be running it and thus commonly auto-fixing issues that scripts cause.

Thanks lethosor for the fast fix (in both senses of the word!)  From the experience I learnt a tiny bit more about the code as well which is always nice.

**Comment by lethosor** [comment]
I don't really have a strong opinion on having it enabled by default, although I'd want to make sure it works well with larger maps first. It could take over 110 MB with a 16x16x200 map, for example, and I could decrease that to around 15 MB just by using individual bits instead of bytes.

**Comment by Putnam3145** [comment]
There is an extant bug in 0.40.24 (fixed in the upcoming version) that messes up unit occupancy in certain situations involving fluid flows. I would discourage running this repeatedly with slowdown, but it's not a DFHack-exclusive condition.

**Comment by lethosor** [comment]
Should be fixed in r5.
